Source/CarthageKit/Project.swift

```diff
@@ -591,6 +591,34 @@ public final class Project { // swiftlint:disable:this type_body_length
 			.flatMap(.concat, unarchive(archive:))
 			.flatMap(.concat) { directoryURL -> SignalProducer<URL, CarthageError> in
 				return frameworksInDirectory(directoryURL)
+					.collect()
+					// Check if the archive contains multiple framewroks resolving to the same unique destiation path.
+					// This check is needed because archives will be copied over to the same destination.
+					.flatMap(.merge) { frameworksUrls -> SignalProducer<URL, CarthageError> in
+						let urlProducer = SignalProducer<URL, CarthageError>(frameworksUrls)
+						return urlProducer
+							.flatMap(.merge) { url -> SignalProducer<String, CarthageError> in
+								return platformForFramework(url)
+									.attemptMap { platform -> Result<String, CarthageError> in
+										guard let path = platform
+											.relativeURL?
+											.appendingPathComponent(url.lastPathComponent, isDirectory: true)
+											.path else {
+												return .failure(.internalError(description: "failed to construct archive destination path from \(platform) and \(url.lastPathComponent)"))
+										}
+										return .success(path)
+								}
+							}
+							.uniqueValues()
+							.collect()
+							.flatMap(.merge) { uniquePaths -> SignalProducer<[String], CarthageError> in
+								guard uniquePaths.count == frameworksUrls.count else {
+									return SignalProducer(error: CarthageError.invalidArchive(description: "found multiple frameworks with the same unarchiving destination"))
+								}
+								return SignalProducer<[String], CarthageError>(value: uniquePaths)
+							}
+							.then(urlProducer)
```

Comment:
This seems like it might be duplicating some of the logic below:

1. Figure out where we would copy everything
2. Make sure there's no duplicates
3. Copy everything
    1. Figure out where we would copy everything
    2. Actually copy it

Can we restructure the code to avoid that duplication?

1. Figure out what to copy where
2. Make sure there are no duplicates
3. Copy based on the list from (1)